src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+                            {
+                                await updateTablesPostInvocationCommand.ExecuteNonQueryAsync(token);
+                            }
+
+                            transaction.Commit();
+                        }
+                        catch (Exception ex)
+                        {
+                            this._logger.LogError("Commit Exception Type: {0}", ex.GetType());
+                            this._logger.LogError("  Message: {0}", ex.Message);
+                            try
+                            {
+                                transaction.Rollback();
+                            }
+                            catch (Exception ex2)
+                            {
+                                this._logger.LogError("Rollback Exception Type: {0}", ex2.GetType());
+                                this._logger.LogError("  Message: {0}", ex2.Message);
+                            }
+                        }
+                    }
+                }
+
+            }
+            catch (Exception e)
+            {
+                // What should we do if releasing the leases fails? We could try to release them again or just wait,
+                // since eventually the lease time will expire. Then another thread will re-process the same changes
+                // though, so less than ideal. But for now that's the functionality.
+                this._logger.LogError($"Failed to release leases for user table {this._userTable.FullName} due to error: {e.Message}");
+            }
+            finally
+            {
+                // Want to do this before releasing the lock in case the renew leases thread wakes up. It will see that
+                // the state is checking for changes and not renew the (just released) leases.
+                await this.ClearRowsAsync(string.Empty, false);
+            }
+        }
+
+        /// <summary>
+        /// Calculates the new version number to attempt to update LastSyncVersion in global state table to. If all
+        /// version numbers in _rows are the same, use that version number. If they aren't, use the second largest
+        /// version number. For an explanation as to why this method was chosen, see 9c in Steps of Operation in this
+        /// design doc: https://microsoft-my.sharepoint.com/:w:/p/t-sotevo/EQdANWq9ZWpKm8e48TdzUwcBGZW07vJmLf8TL_rtEG8ixQ?e=owN2EX.
+        /// </summary>
+        private long RecomputeLastSyncVersion()
+        {
+            var changeVersionSet = new SortedSet<long>();
+            foreach (IReadOnlyDictionary<string, string> row in this._rows)
+            {
+                string changeVersion = row["SYS_CHANGE_VERSION"];
+                changeVersionSet.Add(long.Parse(changeVersion, CultureInfo.InvariantCulture));
+            }
+
+            // If there are at least two version numbers in this set, return the second highest one. Otherwise, return
```

Comment:
Why the second highest?

Reply:
It's documented in the design doc. There must still be more pending items with the highest version, so the second highest version is safe one for cleanup. Though there's an interesting bug noted in the design document for not updating the global version to the highest one, when an insert (v1 to v2) and an update (v2 to v3) in rare case might get treated as two inserts (v1 to v2 and v1 to v3).

Reply:
Please add info about why this is done to the comment - we want to be able to know why unintuitive stuff like this was done without having to dig up the design doc for everything.

Reply:
Moving this inside #282. I will add comment once I completely understand the intricacies of last sync version selection and associated behaviors.